oemof/solph/custom.py

```diff
@@ -839,3 +844,593 @@ def tes_ub_constr_rule(block, n, t):
+
+    Parameters
+    ----------
+    demand: numeric
+        original electrical demand
+    capacity_up: int or array
+        maximum DSM capacity that may be increased
+    capacity_down: int or array
+        maximum DSM capacity that may be reduced
+    method: 'interval' , 'delay'
+
+        interval :
+
+            Simple model in which the load shift must be compensated in a
+            predefined fixed interval(**shift_interval is mandatory).
+            Foundation of this optimisation should be a potential analysis
+            of the DSM capacity during each interval. With this,
+            the boundaries for the DSM variable are set.
+
+        delay :
+
+            Sophisticated model based on the formulation by
+            Zerrahn & Schill (DIW).T he load-shift of the component must be
+            compensated in a predefined delay-time (**delay_time is mandatory).
+            DSM capacity can either be a fixed value or an hourly time series.
+    **shift_interval: int
+
+        only mandatory in method = 'interval' :
+
+        interval in between which total DSM  must be fully compensated
+    **delay_time: int
+
+        only mandatory in method =  'delay':
+
+        the time after which the load shift must be fully compensated
+
+    Notes
+    -----
+
+    The following sets, variables, constraints and objective parts are created
+    :py:class:`~oemof.solph.custom.SinkDSMIntervalBlock` or
+    :py:class:`~oemof.solph.custom.SinkDSMDelayBlock
+    * This component is still under development.
+
+    * As many constraints and dependencies are created in method 'delay',
+    the calculation time increases significantly with large 'delay_time'
+    values as well as with long observation periods.
+
+    * The interval method creates only one real variable,
+    while the delay method creates two non-negative-real variables.
+
+    * The interval always starts at the first timestep = 0.
+
+    * delay method based on the formulation given in a paper
+    by Zerrahn, Alexander and Schill, Wolf-Peter (2015):
+    On the representation of demand-side management in power system models,
+    in: Energy (84), pp. 840-845, 10.1016/j.energy.2015.03.037,
+    accessed 17.09.2019, pp. 842-843.
+
+    * in 'delay' method: delay time might be extended artificially
+    if DSM-capacity isn't used to it's full extend.
+    see oemof/Issue: #622 for more information
+
+    """
+
+    def __init__(self, demand, capacity_up, capacity_down, method,
+                 *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.capacity_up = sequence(capacity_up)
+        self.capacity_down = sequence(capacity_down)
+        self.demand = sequence(demand)
+        self.method = method
+        self.shift_interval = kwargs.get('shift_interval', None)
+        self.delay_time = kwargs.get('delay_time', None)
+        self.recovery_time = kwargs.get('recovery_time', False)
```

Comment:
What is recovery time for? I do not find it anywhere else in the code or the docs.

Reply:
a not used parameter, from and additional not working constraint I already removed. removed it.